pkg/controller/webhookbootstrap/controller.go

```diff
@@ -0,0 +1,488 @@
+	log := logf.FromContext(ctx)
+	log = log.WithValues(logf.ResourceNameKey, name, logf.ResourceNamespaceKey, c.webhookNamespace, logf.ResourceKindKey, "Secret")
+	_, err := c.secretLister.Secrets(c.webhookNamespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Info("Existing Secret does not exist, creating new empty secret")
+		c.createEmptySecret(ctx, log, name)
+		return
+	}
+	if err != nil {
+		log.Error(err, "Failed to GET existing Secret resource")
+		return
+	}
+}
+
+func (c *controller) createEmptySecret(ctx context.Context, log logr.Logger, name string) {
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: c.webhookNamespace,
+			Annotations: map[string]string{
+				cmapi.AllowsInjectionFromSecretAnnotation: "true",
+			},
+		},
+		Data: map[string][]byte{
+			corev1.TLSCertKey:       nil,
+			corev1.TLSPrivateKeyKey: nil,
+			cmapi.TLSCAKey:          nil,
+		},
+		Type: corev1.SecretTypeTLS,
+	}
+	if _, err := c.kubeClient.CoreV1().Secrets(c.webhookNamespace).Create(s); err != nil {
+		log.Error(err, "Failed to create new empty Secret")
+	}
+	return
+}
+
+const (
+	defaultSelfSignedIssuerName = "cert-manager-webhook-selfsigner"
+	defaultCAIssuerName         = "cert-manager-webhook-ca"
+
+	defaultCAKeyAlgorithm = cmapi.RSAKeyAlgorithm
+	defaultCAKeySize      = 2048
+	defaultCAKeyEncoding  = cmapi.PKCS1
+
+	defaultServingKeyAlgorithm = cmapi.RSAKeyAlgorithm
+	defaultServingKeySize      = 2048
+	defaultServingKeyEncoding  = cmapi.PKCS1
+)
+
+func buildCACertificate(secret *corev1.Secret) *cmapi.Certificate {
+	return &cmapi.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            secret.Name,
+			Namespace:       secret.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(secret, corev1.SchemeGroupVersion.WithKind("Secret"))},
+		},
+		Spec: cmapi.CertificateSpec{
+			SecretName:   secret.Name,
+			Organization: []string{"cert-manager.system"},
+			CommonName:   "cert-manager.webhook.ca",
+			// root CA is valid for 5 years as we don't currently handle
+			// rotating the root properly
+			Duration: &metav1.Duration{Duration: time.Hour * 24 * 365 * 5},
+			IssuerRef: cmapi.ObjectReference{
+				Name: defaultSelfSignedIssuerName,
+			},
+			IsCA:         true,
+			KeyAlgorithm: defaultCAKeyAlgorithm,
+			KeySize:      defaultCAKeySize,
+			KeyEncoding:  defaultCAKeyEncoding,
+		},
+	}
+}
+
+func buildServingCertificate(secret *corev1.Secret, dnsNames []string) *cmapi.Certificate {
+	return &cmapi.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            secret.Name,
+			Namespace:       secret.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(secret, corev1.SchemeGroupVersion.WithKind("Secret"))},
+		},
+		Spec: cmapi.CertificateSpec{
+			SecretName:   secret.Name,
+			Organization: []string{"cert-manager.system"},
+			DNSNames:     dnsNames,
+			Duration:     &metav1.Duration{Duration: time.Hour * 24 * 365 * 1},
```

Comment:
Do we definitely want to have a year here over say 90 days?

Reply:
I'm inclined to say yes, but I'm open to changing it...